### ISSUE TITLE
Prefetch to device with mesh and PartitionSpec

### ISSUE THREAD

**Issue by gianlucadetommaso** [google/flax#3178]
[prefetch_to_device](https://flax.readthedocs.io/en/latest/api_reference/flax.jax_utils.html#flax.jax_utils.prefetch_to_device) works well with `pmap` to prefetch on multiple devices. However I think it assumes that each batch in the data loader will be distributed across a certain axis.

In my case, I would like to prefetch data when the batch has been put to device using a certain mesh and a `PartitionSpec`. Given a `mesh` with axes `('x', 'y', 'z')`, and a certain data loader `loader`, my code may look like this:

```
from jax.sharding import NamedSharding, PartitionSpec

def _shard(batch):
      return device_put(batch, NamedSharding(mesh, PartitionSpec(("x", "y"))))
      
loader = map(lambda batch: tree_map(_shard, batch), loader)           
```

At this point, I would like to prefetch to device using

```
loader = jax_utils.prefetch_to_device(loader, 2)
```

but this doesn't work because the first dimension of the `batch` does not match the number of devices. How can I prefetch to the devices specified by my mesh and PartitionSpec?

Thanks!

**Comment by cgarciae** [comment]
I haven't tested this code, just modifying `prefectch_to_device` a bit to accept a `mesh` and a the spec of the data you can build something like e.g. `prefetch_to_mesh`:

```python
import collections
import itertools
import jax
from jax.sharding import NamedSharding, Mesh

def prefetch_to_mesh(iterator, size: int, mesh: Mesh, xs_spec):
  queue = collections.deque()

  def _prefetch(xs):
    return jax.device_put(xs, NamedSharding(mesh, xs_spec))

  def enqueue(n):  # Enqueues *up to* `n` elements from the iterator.
    for data in itertools.islice(iterator, n):
      queue.append(jax.tree_util.tree_map(_prefetch, data))

  enqueue(size)  # Fill up the buffer.
  while queue:
    yield queue.popleft()
    enqueue(1)
```

**Comment by gianlucadetommaso** [comment]
Seems to be doing what's supposed to, thanks a lot!
